app/helpers/automated_tests_client_helper.rb

```diff
@@ -428,24 +392,121 @@ def self.perform(host_with_port, test_scripts, user_api_key, server_api_key, gro
     end
   end
 
-  def self.process_test_result(raw_result, test_scripts_ran, assignment, grouping, submission, requested_by)
+  def self.process_test_result(xml, test_script_result)
+
+    test_name = xml['name']
+    if test_name.nil?
+      add_test_error_result(test_script_result, I18n.t('automated_tests.test_result.unknown_test'),
+                            I18n.t('automated_tests.test_result.bad_results', {xml: xml}))
+      raise 'Malformed xml'
+    end
+
+    marks_earned = xml['marks_earned'].nil? ? 0.0 : xml['marks_earned'].to_f
+    marks_total = xml['marks_total'].nil? ? nil : xml['marks_total'].to_f
+    input = xml['input'].nil? ? '' : xml['input']
+    expected = xml['expected'].nil? ? '' : xml['expected']
+    actual = xml['actual'].nil? ? '' : xml['actual']
+    status = xml['status']
+    if status.nil? or not status.in?(%w(pass partial fail error error_all))
+      actual = I18n.t('automated_tests.test_result.bad_status', {status: status})
+      status = 'error'
+      marks_earned = 0.0
+    end
+    if status == 'error_all'
+      status = 'error'
+      stop_processing = true
+    else
+      stop_processing = false
+    end
+    if !marks_total.nil? && marks_earned > marks_total
+      marks_earned = marks_total
+    end
+
+    add_test_result(test_script_result, test_name, input, actual, expected, marks_earned, marks_total, status)
+    if stop_processing
+      raise 'Test script reported a serious failure'
+    end
+
+    return marks_earned, marks_total
+  end
 
+  def self.process_test_script_result(xml, assignment, grouping, submission, requested_by)

```

Comment:
As above, I could see this moving this method to something like `TestScriptResult.create_from_xml`.